src/micro.c

```diff
@@ -0,0 +1,680 @@
+    info->Version = m->cfg->Version;
+    info->Description = m->cfg->Description;
+    info->Id = m->id;
+    info->Type = MICRO_INFO_RESPONSE_TYPE;
+
+    // Overallocate subjects, will filter out internal ones.
+    info->Subjects = NATS_CALLOC(m->num_eps, sizeof(char *));
+    if (info->Subjects == NULL)
+    {
+        NATS_FREE(info);
+        return micro_ErrorOutOfMemory;
+    }
+
+    micro_lock_service(m);
+    for (ep = m->first_ep; ep != NULL; ep = ep->next)
+    {
+        if ((!ep->is_monitoring_endpoint) && (ep->subject != NULL))
+        {
+            info->Subjects[len] = ep->subject;
+            len++;
+        }
+    }
+    info->SubjectsLen = len;
+    micro_unlock_service(m);
+
+    *new_info = info;
+    return NULL;
+}
+
+void microServiceInfo_Destroy(microServiceInfo *info)
+{
+    if (info == NULL)
+        return;
+
+    // subjects themselves must not be freed, just the collection.
+    NATS_FREE(info->Subjects);
+    NATS_FREE(info);
+}
+
+microError *
+microService_GetStats(microServiceStats **new_stats, microService *m)
+{
+    microServiceStats *stats = NULL;
+    microEndpoint *ep = NULL;
+    int len = 0;
+    long double avg = 0.0;
+
+    if ((new_stats == NULL) || (m == NULL) || (m->service_mu == NULL))
+        return micro_ErrorInvalidArg;
+
+    stats = NATS_CALLOC(1, sizeof(microServiceStats));
+    if (stats == NULL)
+        return micro_ErrorOutOfMemory;
+
+    stats->Name = m->cfg->Name;
+    stats->Version = m->cfg->Version;
+    stats->Id = m->id;
+    stats->Started = m->started;
+    stats->Type = MICRO_STATS_RESPONSE_TYPE;
+
+    // Allocate the actual structs, not pointers. Overallocate for the internal
+    // endpoints even though they are filtered out.
+    stats->Endpoints = NATS_CALLOC(m->num_eps, sizeof(microEndpointStats));
+    if (stats->Endpoints == NULL)
+    {
+        NATS_FREE(stats);
+        return micro_ErrorOutOfMemory;
+    }
+
+    micro_lock_service(m);
+    for (ep = m->first_ep; ep != NULL; ep = ep->next)
+    {
+        if ((ep != NULL) && (!ep->is_monitoring_endpoint) && (ep->endpoint_mu != NULL))
+        {
+            micro_lock_endpoint(ep);
+            // copy the entire struct, including the last error buffer.
+            stats->Endpoints[len] = ep->stats;
+
+            stats->Endpoints[len].Name = ep->name;
+            stats->Endpoints[len].Subject = ep->subject;
+            avg = (long double)ep->stats.processing_time_s * 1000000000.0 + (long double)ep->stats.processing_time_ns;
+            avg = avg / (long double)ep->stats.num_requests;
+            stats->Endpoints[len].average_processing_time_ns = (int64_t)avg;
+            len++;
+            micro_unlock_endpoint(ep);
+        }
+    }
+    micro_unlock_service(m);
+    stats->EndpointsLen = len;
+
+    *new_stats = stats;
+    return NULL;
+}
+
+void microServiceStats_Destroy(microServiceStats *stats)
+{
+    if (stats == NULL)
+        return;
+
+    NATS_FREE(stats->Endpoints);
```

Comment:
`Endpoints` is a struct and I see has strings, but they are `const` and point to the endpoint's own string, so that is why there is no need to have a dedicated destroy for that struct, correct?

Reply:
The same consideration applies here as to `Subjects` in the info case, I think. `Name, Subject` are indeed references to the endpoint itself, and are returned as `const char *`. It still may be wiser to dup/free them. No other fields in `microEndpointStats` are allocated separately.

Reply:
And see my comment I made there: https://github.com/nats-io/nats.c/pull/640#discussion_r1192667971

Reply:
Duplicated them all.